### PR TITLE
make sure that ideals in graded rings of type `MPolyRing`and `MPolyQu…

### PR DESCRIPTION
**Pull request by wdecker**
…oRing`are given by homogeneous generators

### CODE REVIEW

**Review by thofma — APPROVED** [review]

**Comment by wdecker** [comment]
@HereAround, @lkastner: This test runs for quite some time before it conflicts with allowing only ideals given by homogeneous generators if we are in graded situations:

julia> K, a = cyclotomic_field(4)
(Cyclotomic field of order 4, z_4)

julia> G = matrix_group(
              matrix(K, 4, 4, [ 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0 ]),
              matrix(K, 4, 4, [ a, 0, 0, 0, 0, inv(a), 0, 0, 0, 0, inv(a), 0, 0, 0, 0, a ])
             ) # symplectic reflection representation of dihedral group of order 8
Matrix group of degree 4 over K

julia> L = linear_quotient(G)
Linear quotient by Matrix group of degree 4 over K

julia> C, CtoRt = Oscar.cox_ring_of_qq_factorial_terminalization(L)
ERROR: ArgumentError: The generators of the ideal must be homogeneous.
Stacktrace:
 [1] MPolyIdeal(B::Oscar.IdealGens{MPolyDecRingElem{nf_elem, AbstractAlgebra.Generic.MPoly{nf_elem}}})
   @ Oscar ~/.julia/dev/Oscar/src/Rings/mpoly.jl:598
 [2] MPolyIdeal
   @ ~/.julia/dev/Oscar/src/Rings/mpoly.jl:577 [inlined]
 [3] ideal(R::MPolyDecRing{nf_elem, AbstractAlgebra.Generic.MPolyRing{nf_elem}}, g::Vector{AbstractAlgebra.Generic.MPoly{nf_elem}})
   @ Oscar ~/.julia/dev/Oscar/src/Rings/mpoly-ideals.jl:39
 [4] cox_ring(L::Oscar.LinearQuotient{nf_elem, AbstractAlgebra.Generic.MatSpaceElem{nf_elem}}; algo_gens::Symbol, algo_rels::Symbol)
   @ Oscar ~/.julia/dev/Oscar/experimental/LinearQuotients/src/cox_rings.jl:304
 [5] cox_ring
   @ ~/.julia/dev/Oscar/experimental/LinearQuotients/src/cox_rings.jl:242 [inlined]
 [6] cox_ring_of_qq_factorial_terminalization(L::Oscar.LinearQuotient{nf_elem, AbstractAlgebra.Generic.MatSpaceElem{nf_elem}})
   @ Oscar ~/.julia/dev/Oscar/experimental/LinearQuotients/src/cox_rings.jl:333
 [7] top-level scope
   @ REPL[17]:1

**Comment by thofma** [comment]
I think this is functionality added by @joschmitt

**Comment by joschmitt** [comment]
Yes, I know what the problem is and will look into it tomorrow.

**Comment by joschmitt** [comment]
This is not that easy to fix. In the function `cox_ring(::LinearQuotient)` I build an ideal in a polynomial ring graded by a finite abelian group. I know from theory that the ideal is homogeneous with respect to that grading, but I don't know of a way to get homogeneous generators for it. Right now, I don't have time to look into it in more detail. You should probably just disable the 'bad test' since this is experimental functionality anyway.

**Comment by wdecker** [comment]
@joschmitt The fix is easy: If you know that your computed polynomials generate a homogeneous ideal, just replace each such polynomial by its homogeneous components BEFORE creating the ideal in OSCAR.

File: src/Rings/mpoly.jl [hunk 1]
```diff
@@ -589,6 +589,17 @@ Fields:
   end
 
   function MPolyIdeal(B::IdealGens{T}) where T
+    if length(B) >= 1
+      oscar_assure(B)
+      R = B.gens.Ox
+      if R isa MPolyDecRing
+         if is_graded(R)
```
Comment (fingolfin) [inline src/Rings/mpoly.jl:596]: This kind of check could be rewritten to just do `if is_graded(R)
`, which IMHO reads much more naturally, if only we provided methods for `is_graded`, `is_z_graded`, etc. for "regular" rings and ideals where they just return `false`.

I just saw that PR #2236 does this for `is_graded`.

Comment (fingolfin) [reply src/Rings/mpoly.jl:596]: See PR #2241 where I just added such methods

**Comment by joschmitt** [comment]
> @joschmitt The fix is easy: If you know that your computed polynomials generate a homogeneous ideal, just replace each such polynomial by its homogeneous components BEFORE creating the ideal in OSCAR.

Yes, you are right, I'm sorry. I pushed a commit for this to my fork 087849bf039b50df6f5210b4a0ce66432ac57a6e ; maybe someone can cherry-pick it here?
However, now the tests fail due to the following problem:
```
julia> R, (x, y) = graded_polynomial_ring(QQ, ["x", "y"], [ 1, 1 ])
(Multivariate Polynomial Ring in x, y over Rational Field graded by 
  x -> [1]
  y -> [1], MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x, y])

julia> I = ideal(R, [ x*y ])
ideal(x*y)

julia> Q, _ = quo(R, I)
(Quotient of Multivariate Polynomial Ring in x, y over Rational Field graded by 
  x -> [1]
  y -> [1] by ideal(x*y), Map from
R to Q defined by a julia-function with inverse)

julia> intersect(ideal(Q, Q(x)), ideal(Q, Q(y)))
ERROR: MethodError: Cannot `convert` an object of type MPolyQuoRingElem{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}} to an object of type MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}
Closest candidates are:
  convert(::Type{T}, ::T) where T at Base.jl:61
Stacktrace:
  [1] setindex!(A::Vector{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}, x::MPolyQuoRingElem{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}, i1::Int64)
    @ Base ./array.jl:966
  [2] _unsafe_copyto!(dest::Vector{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}, doffs::Int64, src::Vector{MPolyQuoRingElem{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}}, soffs::Int64, n::Int64)
    @ Base ./array.jl:253
  [3] unsafe_copyto!
    @ ./array.jl:307 [inlined]
  [4] _copyto_impl!
    @ ./array.jl:331 [inlined]
  [5] copyto!
    @ ./array.jl:317 [inlined]
  [6] copyto!
    @ ./array.jl:343 [inlined]
  [7] copyto_axcheck!
    @ ./abstractarray.jl:1127 [inlined]
  [8] Array
    @ ./array.jl:626 [inlined]
  [9] convert(#unused#::Type{Vector{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}}, a::Vector{MPolyQuoRingElem{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}})
    @ Base ./array.jl:617
 [10] setproperty!(x::Oscar.BiPolyArray{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}, f::Symbol, v::Vector{MPolyQuoRingElem{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}})
    @ Base ./Base.jl:39
 [11] oscar_assure(B::Oscar.BiPolyArray{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}})
    @ Oscar ~/.julia/dev/Oscar/src/Rings/mpoly.jl:682
 [12] oscar_assure
    @ ~/.julia/dev/Oscar/src/Rings/mpoly.jl:687 [inlined]
 [13] MPolyQuoIdeal(Ox::MPolyQuoRing{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}, si::Singular.sideal{Singular.spoly{Singular.n_Q}})
    @ Oscar ~/.julia/dev/Oscar/src/Rings/MPolyQuo.jl:115
 [14] intersect(a::MPolyQuoIdeal{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}, b::MPolyQuoIdeal{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}})
    @ Oscar ~/.julia/dev/Oscar/src/Rings/MPolyQuo.jl:367
 [15] top-level scope
    @ REPL[17]:1
```
From what I can see, `oscar_assure` for ideals in quotient rings does not work properly. (Note that the tests will only fail if they are included again of course.)

**Comment by wdecker** [comment]
@joschmitt The mistake is in my use of IdealGens here:

function MPolyQuoIdeal(Ox::MPolyQuoRing{T}, si::Singular.sideal) where T <: MPolyRingElem
    singular_quotient_ring(Ox) == base_ring(si) || error("base rings must match")
    B = IdealGens(Ox, si)

I will dig further into it.

**Comment by joschmitt** [comment]
Like this, also the disabled tests work again with my commit 919150430553f085f4090bb43df31a54fdceec9f . Either somebody adds this here or I can open a pull request once this pull request is merged.

**Review by ederc — APPROVED** [review]

**Comment by lkastner** [comment]
> Like this, also the disabled tests work again with my commit [9191504](https://github.com/oscar-system/Oscar.jl/commit/919150430553f085f4090bb43df31a54fdceec9f) . Either somebody adds this here or I can open a pull request once this pull request is merged.

I cherry-picked it here. If someone disagrees feel free to remove.

**Comment by joschmitt** [comment]
> > Like this, also the disabled tests work again with my commit [9191504](https://github.com/oscar-system/Oscar.jl/commit/919150430553f085f4090bb43df31a54fdceec9f) . Either somebody adds this here or I can open a pull request once this pull request is merged.
> 
> I cherry-picked it here. If someone disagrees feel free to remove.

Thank you! Then e80beb7f6caa27c2d5854e7c64535301b8bf11a2 can be reverted.

**Pull request merged**